### ISSUE TITLE
Customise autocorrect

### ISSUE THREAD

**Issue by janapejic** [realm/SwiftLint#350]
Is it possible to customise autocorrect options for a certain rule?

**Comment by scottrhoyt** [comment]
@janapejic I don't believe that behavior is currently supported, though there is some groundwork in place to support more advanced customization like this. Can you give an example of what you would like to customize?

**Comment by janapejic** [comment]
Yeah, sure. Once I integrated SwiftLint, I got a bunch of build errors because of force casting. Now I know that this is bad practise, but in a situation like:
    let cell = collectionView.dequeueReusableCellWithReuseIdentifier("Identifier", forIndexPath: indexPath) as! SomeCell
I do not want to do `if let`, because if it is not possible to cast, my app should not continue running.

My first idea was to make this a warning instead of an error, but as far as I could see such customisation is also not possible. So what I did is I refactored my code to look like this:

    guard let cell = collectionView.dequeueReusableCellWithReuseIdentifier("Identifier", forIndexPath: indexPath) as? SomeCell else { fatalError("Dequeueing SomeCell failed) }

Since my app is rather big, I would have liked to have been abled to autocorrect this :)

**Comment by scottrhoyt** [comment]
I'm not sure this level of customization will be supported soon because it involves managing the different ways that someone might want to correct this problem and keeping them syntactically correct. There is also the context of which the force cast occurs. While you might want to correct your cell dequeuing this way, you may want to handle other force casts differently. The problem gets complex quickly.

My personal preference is that crashing an app should be avoided at most costs. However, if you want to just disable a single force cast, you can use...

    // swiftlint:disable:next force_cast

... on the line before the force cast. On a large project maybe you can use the find/replace project tools to help you track these situations down?

**Comment by janapejic** [comment]
Yes, that is what I did eventually. I do not really like 

    // swiftlint:disable:next force_cast
since it somehow clutters the code and it has to be added everywhere. 

I understand why making such a replacement as default could create problems. that's why I still think it would be good to allow for more customisability so that I can by myself define how I want to autocorrect something.

**Comment by scottrhoyt** [comment]
If you are finding that you are putting that everywhere, then potentially that is a code smell that another strategy can be used, since that implies that there are a lot of places where the code can crash. Personally, I think `guard let`, `if let`, or optional chaining should be used to mitigate this. 

However, in the case where you want to crash hard when cells can't be dequeued correctly, you could use another strategy like subclassing or extending `UITableViewController` to add a generic method that will force cast to a specific subclass of `UITableViewCell` or crash if that's not possible. Then you should only have to deal with this in one place. Here is a playground snippet to illustrate:

``` swift
import UIKit

extension UITableView {
    func forceDequeueCell<T: UITableViewCell>(identifier: String) -> T {
        // swiftlint:disable:next force_cast
        return dequeueReusableCellWithIdentifier(identifier) as! T
    }
}

class MyCell: UITableViewCell {}

let tableView = UITableView()

let myCell1: MyCell = tableView.forceDequeueCell("MyCell") // Crash

tableView.registerClass(MyCell.self, forCellReuseIdentifier: "MyCell")

let myCell2: MyCell = tableView.forceDequeueCell("MyCell") // Works
```

**Comment by scottrhoyt** [comment]
Closing this because of the potential solutions listed above and not being able to support this level of customization in `autocorrect`. This seems it would be handled better on a case by case basis or perhaps as a analyzer or compiler features as opposed to a linter. Thanks.
